pkg/utils/report/results.go

```diff
@@ -86,72 +87,76 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
-				}
-			}
-			pss := ruleResult.PodSecurityChecks()
-			if pss != nil {
-				var controls []string
-				for _, check := range pss.Checks {
-					if !check.CheckResult.Allowed {
-						controls = append(controls, check.ID)
-					}
-				}
-				if len(controls) > 0 {
-					sort.Strings(controls)
-					result.Properties = map[string]string{
-						"standard": string(pss.Level),
-						"version":  pss.Version,
-						"controls": strings.Join(controls, ","),
-					}
-				}
-			}
-			if result.Result == "fail" && !result.Scored {
-				result.Result = "warn"
-			}
-			results = append(results, result)
+func addProperty(k, v string, result *policyreportv1alpha2.PolicyReportResult) {
+	if result.Properties == nil {
+		result.Properties = map[string]string{}
+	}
+
+	result.Properties[k] = v
+}
+
+func ToPolicyReportResult(policyType engineapi.PolicyType, policyName string, ruleResult engineapi.RuleResponse, annotations map[string]string, resource *corev1.ObjectReference) policyreportv1alpha2.PolicyReportResult {
+	result := policyreportv1alpha2.PolicyReportResult{
+		Source:  kyverno.ValueKyvernoApp,
+		Policy:  policyName,
+		Rule:    ruleResult.Name(),
+		Message: ruleResult.Message(),
+		Result:  toPolicyResult(ruleResult.Status()),
+		Scored:  annotations[kyverno.AnnotationPolicyScored] != "false",
+		Timestamp: metav1.Timestamp{
+			Seconds: time.Now().Unix(),
+		},
+		Category: annotations[kyverno.AnnotationPolicyCategory],
+		Severity: SeverityFromString(annotations[kyverno.AnnotationPolicySeverity]),
+	}
+	if result.Result == "fail" && !result.Scored {
+		result.Result = "warn"
+	}
+	if resource != nil {
+		result.Resources = []corev1.ObjectReference{
+			*resource,
 		}
-	} else {
-		for _, ruleResult := range response.PolicyResponse.Rules {
-			result := policyreportv1alpha2.PolicyReportResult{
-				Source:  "ValidatingAdmissionPolicy",
-				Policy:  ruleResult.Name(),
-				Message: ruleResult.Message(),
-				Result:  toPolicyResult(ruleResult.Status()),
-				Timestamp: metav1.Timestamp{
-					Seconds: time.Now().Unix(),
-				},
-			}
-			if ruleResult.ValidatingAdmissionPolicyBinding() != nil {
-				result.Properties = map[string]string{
-					"binding": ruleResult.ValidatingAdmissionPolicyBinding().Name,
-				}
+	}
+	if ruleResult.Exception() != nil {
+		addProperty("exception", ruleResult.Exception().Name, &result)
+	}
+	pss := ruleResult.PodSecurityChecks()
+	if pss != nil && len(pss.Checks) > 0 {
+		var controls []string
+		for _, check := range pss.Checks {
+			if !check.CheckResult.Allowed {
+				controls = append(controls, check.ID)
 			}
-			results = append(results, result)
+		}
+		if len(controls) > 0 {
+			sort.Strings(controls)
+			addProperty("standard", string(pss.Level), &result)
+			addProperty("version", pss.Version, &result)
+			addProperty("controls", strings.Join(controls, ","), &result)
+		}
+	}
+	if policyType == engineapi.ValidatingAdmissionPolicyType {
+		result.Source = "ValidatingAdmissionPolicy"
```

Comment:
@JimBugwadia category and severity field are added here, which we don't need in case of VAP, I think that's the reason why one of the chainsaw test is still failing.

Reply:
@coder12git - will take a look. 

We do want those, if they are declared in the ValidatingAdmissionPolicy, as annotations.

Reply:
That's the reason why i added if and else statement, to remove those fields in case of VAP.